File: Foundation/Scratch/String/ques1.cpp
// ques:
// input: abaabbbccddeffghhh
// output: a1b1a2b3c2d2e1f2g1h3

#include<iostream>
using namespace std;
string stringCompress(string str){
        if(str.length()==0) return "";
        string sb = "";
        char prevChar = str[0];
        int i = 1;
        while(i <= str.length()){
            int count = 1;
            while(i < str.length() && prevChar == str[i]){
                count++;
                prevChar = str[i++];
            }
            sb.push_back(prevChar);
            sb += to_string(count);
            if(i==str.length())
               break;
            prevChar = str[i++];
        }
        return sb;
    }
int main(){
    string str = "abaabbbccddeffghhh";
    cout<<stringCompress(str);
    return 0;
}

File: Foundation/Scratch/String/ques1.java
// ques:
// input: abaabbbccddeffghhh
// output: a1b1a2b3c2d2e1f2g1h3

public class ques1{
    public static String stringCompress(String str){
        if(str.length()==0) return "";
        StringBuilder sb = new StringBuilder();
        char prevChar = str.charAt(0);
        int i = 1;
        while(i <= str.length()){
            int count = 1;
            while(i < str.length() && prevChar == str.charAt(i)){
                count++;
                prevChar = str.charAt(i++);
            }
            sb.append(prevChar);
            sb.append(count);
            if(i==str.length())
               break;
            prevChar = str.charAt(i++);
        }
        return sb.toString();
    }
    public static void main(String[] args){
        String str = "abaabbbccddeffghhh";
        System.out.println(stringCompress(str));
    }
}
